src/PubSub/ResourceNameTrait.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/**
+ * Copyright 2016 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\PubSub;
+
+trait ResourceNameTrait
+{
+    /**
+     * @var array
+     */
+    private $templates = [
+        'project' => 'projects/%1$s',
+        'topic' => 'projects/%2$s/topics/%1$s',
+        'subscription' => 'projects/%2$s/subscriptions/%1$s'
+    ];
+
+    /**
+     * @var array
+     */
+    private $regexes = [
+        'project' => '/^projects\/([^\/]*)$/',
+        'topic' => '/projects\/[^\/]*\/topics\/(.*)/',
+        'subscription' => '/projects\/[^\/]*\/subscriptions\/(.*)/'
+    ];
+
+    /**
+     * Convert a fully-qualified name into a simple name.
+     *
+     * ```php
+     * <?php
+     * use Google\Cloud\ServiceBuilder;
+     *
+     * $cloud = new ServiceBuilder([
+     *      'projectId' => 'my-awesome-project'
+     * ]);
+     *
+     * $pubsub = $cloud->pubsub();
+     * $topic = $pubsub->topic('projects/my-awesome-project/topics/my-topic-name');
+     * echo $topic->pluckName($name); // `my-topic-name`
+     * ```
+     *
+     * @param  string $name
+     * @return string
+     */
+    public function pluckName($type, $name)
+    {
+        if (!isset($this->regexes[$type])) {
+            throw new \InvalidArgumentException(sprintf(
+                'Regex `%s` is not defined',
+                $type
+            ));
+        }
+
+        $matches = [];
+        $res = preg_match($this->regexes[$type], $name, $matches);
+        return ($res === 1) ? $matches[1] : null;
+    }
+
+    /**
+     * Convert a simple name into the fully-qualified name required by
+     * the API.
+     *
+     * ```php
+     * <?php
+     * use Google\Cloud\ServiceBuilder;
+     *
+     * $cloud = new ServiceBuilder([
+     *      'projectId' => 'my-awesome-project'
+     * ]);
+     *
+     * $pubsub = $cloud->pubsub();
+     * $topic = $pubsub->topic('my-topic-name');
+     * echo $topic->formatName($name); // `projects/my-awesome-project/topics/my-topic-name`
+     * ```
+     *
+     * @param  string $type
+     * @param  array  $pieces {
+     *      Template Values
+     *
+     *      [0] @type string The simple name of the resource. Could be a project ID
```

Comment:
This function expects a non-associative array:

    ['topicName', 'projectId']

Is there a better way to document the required values?